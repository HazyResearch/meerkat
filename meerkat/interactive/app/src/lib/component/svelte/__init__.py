from meerkat.interactive.app.src.lib.component.abstract import Component, Slottable
from meerkat.mixins.identifiable import classproperty


class SvelteMixin:
    @classproperty
    def library(cls):
        return "svelte"

    @classproperty
    def namespace(cls):
        return "svelte"

# class Lettable:

#     def __init__(self, let: dict = None, *args, **kwargs):
#         super().__init__(*args, **kwargs)
#         self.let = let

# class fragment(Slottable, SvelteMixin, Component):

#     slot: str = None


# class options(Slottable, SvelteMixin, Component):

#     immutable: bool = False
#     accessors: bool = False
#     namespace: str = None
#     tag: str = None


# class head(Slottable, SvelteMixin, Component):

#     pass


# class component(Slottable, SvelteMixin, Component):

#     this: str = None
#     props: dict = None


# class self(Slottable, SvelteMixin, Component):

#     props: dict = None


# class each(Slottable, Component):

#     items: list = None
