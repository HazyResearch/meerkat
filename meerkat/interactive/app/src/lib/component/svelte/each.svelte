<script>
    export let items;
</script>

{#each items as item}
    <slot {item} />
{/each}