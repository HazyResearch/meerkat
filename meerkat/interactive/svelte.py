import importlib.util
import logging
import os
import shutil
import sys
from typing import TYPE_CHECKING, List, Set, Type

from tabulate import tabulate

from meerkat.constants import (
    JINJA_ENV,
    MEERKAT_INIT_PROCESS,
    MEERKAT_NPM_PACKAGE,
    MEERKAT_RUN_ID,
    MEERKAT_RUN_PROCESS,
    MEERKAT_RUN_RELOAD_COUNT,
    MEERKAT_RUN_SUBPROCESS,
    App,
    PathHelper,
    write_file,
)
from meerkat.interactive import BaseComponent
from meerkat.tools.filelock import FileLock
from meerkat.tools.singleton import Singleton

if TYPE_CHECKING:
    from meerkat.interactive import Page

logger = logging.getLogger(__name__)


def get_subclasses_recursive(cls: type) -> List[type]:
    """Recursively find all subclasses of a class.

    Args:
        cls (type): the class to find subclasses of.

    Returns:
        List[type]: a list of all subclasses of cls.
    """
    subclasses = []
    for subclass in cls.__subclasses__():
        subclasses.append(subclass)
        subclasses.extend(get_subclasses_recursive(subclass))
    return subclasses


def write_file_if_changed(path: str, content: str):
    """Write a file if the content has changed. Note this is not atomic.

    Args:
        path (str): the path to write to.
        content (str): the content to write.
    """
    if os.path.exists(path):
        with open(path, "r") as f:
            if f.read() == content:
                return
    write_file(path, content)


class SvelteWriter(metaclass=Singleton):
    """Class that handles writing Svelte components to a Meerkat app."""

    def __init__(self):
        self.app = App(appdir=PathHelper().appdir)

        self._ran_import_app_components = False
        self._components = None
        self._frontend_components = None

    @property
    def appdir(self):
        return self.app.appdir

    def run(self):
        """Write component wrappers and context at the start of a run."""
        self.import_app_components()
        with FileLock(os.path.join(self.appdir, "svelte_writer")):
            self.cleanup()
            self.write_all_component_wrappers()  # src/lib/wrappers/
            self.write_component_context()  # ComponentContext.svelte

    def cleanup(self):
        """Cleanup the app."""
        self.remove_all_component_wrappers()
        self.remove_component_context()
        logger.debug("Removed all component wrappers and ComponentContext.svelte.")

    def get_all_components(
        self,
        exclude_classes: Set[str] = {"Component", "BaseComponent"},
    ) -> List[Type["BaseComponent"]]:
        """Get all subclasses of BaseComponent, excluding the ones in
        `exclude_classes`.

        Args:
            exclude_classes (Set[str], optional): Set of classes
                to exclude. Defaults to {"Component", "BaseComponent"}.

        Returns:
            List[Type["BaseComponent"]]: List of subclasses of BaseComponent.
        """
        if self._components:
            return self._components

        # Recursively find all subclasses of Component
        subclasses = get_subclasses_recursive(BaseComponent)

        # Filter out the classes we don't want and sort
        subclasses = [c for c in subclasses if c.__name__ not in exclude_classes]
        subclasses = sorted(subclasses, key=lambda c: c.alias)

        tabulated_subclasses = tabulate(
            [[subclass.__module__, subclass.__name__] for subclass in subclasses]
        )
        logger.debug(f"Found {len(subclasses)} components.\n" f"{tabulated_subclasses}")

        self._components = subclasses
        return subclasses

    def get_all_frontend_components(self) -> List[Type["BaseComponent"]]:
        """Get all subclasses of BaseComponent that have a unique
        frontend_alias.

        Returns:
            List[Type["BaseComponent"]]: List of subclasses of BaseComponent.
        """
        if self._frontend_components:
            return self._frontend_components

        # Create a `frontend_components` list that contains the
        # components that have unique component.frontend_alias
        components = self.get_all_components()
        frontend_components = []
        aliases = set()
        for component in components:
            if component.frontend_alias not in aliases:
                frontend_components.append(component)
                aliases.add(component.frontend_alias)

        self._frontend_components = frontend_components
        return frontend_components

    def import_app_components(self):
        """Import all components inside the app/src/lib/components directory to
        register custom user components from their app."""
        if self._ran_import_app_components:
            # Only run this once in a process
            return

        if self.app.is_user_app:
            # Import all components inside the app/src/lib/components
            # directory to register user components from the app
            # Otherwise do nothing
            logger.debug(
                "The app being run is as a user app. "
                f"Adding {self.appdir} to sys.path. "
                "Importing app components from app/src/lib/components."
            )

            # StackOverflow:
            # How can I import a module dynamically given the full path?
            # https://stackoverflow.com/a/67692
            # This module name can be anything
            module_name = "user.app.src.lib.components"
            spec = importlib.util.spec_from_file_location(
                module_name,
                f"{self.appdir}/src/lib/components/__init__.py",
            )
            module = importlib.util.module_from_spec(spec)
            sys.modules[module_name] = module
            spec.loader.exec_module(module)

        self._ran_import_app_components = True

    def remove_all_component_wrappers(self):
        """Remove all component wrappers from the app."""
        try:
            shutil.rmtree(f"{self.appdir}/src/lib/wrappers")
        except (FileNotFoundError, OSError):
            pass

    def remove_component_context(self):
        """Remove the ComponentContext.svelte file from the app."""
        try:
            os.remove(f"{self.appdir}/src/lib/ComponentContext.svelte")
        except OSError:
            pass

    def render_component_context(self):
        """Render the ComponentContext.svelte file for the app."""
        template = JINJA_ENV.get_template("ComponentContext.svelte")
        components = self.get_all_components()
        frontend_components = self.get_all_frontend_components()

        # Get the libraries that the frontend components
        # depend on
        libraries = set([c.library for c in frontend_components])

        # Filter to only include components and frontend components
        # whose libraries are installed
        installed_libraries = self.app.filter_installed_libraries(libraries) + ["html"]
        components = [
            c
            for c in components
            if c.library in installed_libraries
            or c.library == MEERKAT_NPM_PACKAGE
            and not self.app.is_user_app
        ]
        frontend_components = [
            c
            for c in frontend_components
            if c.library in installed_libraries
            or c.library == MEERKAT_NPM_PACKAGE
            and not self.app.is_user_app
        ]

        # For the Meerkat npm package, check the components offered by the
        # user's installed version, and filter out the ones that aren't available
        if MEERKAT_NPM_PACKAGE in installed_libraries and self.app.is_user_app:
            try:
                mk_components = set(
                    [f"Meerkat{c}" for c in self.app.get_mk_package_info()]
                )
                components = [
                    c
                    for c in components
                    if (c.frontend_alias in mk_components and c.namespace == "meerkat")
                    or (c.library == MEERKAT_NPM_PACKAGE and c.namespace != "meerkat")
                    or c.library != MEERKAT_NPM_PACKAGE
                ]
                frontend_components = [
                    c
                    for c in frontend_components
                    if (c.frontend_alias in mk_components and c.namespace == "meerkat")
                    or (c.library == MEERKAT_NPM_PACKAGE and c.namespace != "meerkat")
                    or c.library != MEERKAT_NPM_PACKAGE
                ]
            except Exception as e:
                logger.error(
                    "Error getting Meerkat package info. "
                    "Components from the Meerkat npm package may not be available."
                )
                logger.debug(e)

        return template.render(
            components=components,
            frontend_components=frontend_components,
        )

    def render_component_wrapper(self, component: Type[BaseComponent]):
        # TODO: fix line breaks in Wrapper.svelte
        template = JINJA_ENV.get_template("Wrapper.svelte")
        from meerkat.interactive.startup import snake_case_to_camel_case

        prop_names_camel_case = [
            snake_case_to_camel_case(prop_name) for prop_name in component.prop_names
        ]

        return template.render(
            import_style=component.wrapper_import_style,
            component_name=component.component_name,
            path=component.path,
            prop_names=component.prop_names,
            prop_names_camel_case=prop_names_camel_case,
            event_names=component.event_names,
            use_bindings=True,
            prop_bindings=component.prop_bindings,
            slottable=component.slottable,
            zip=zip,
            is_user_app=self.app.is_user_app,
        )

    def get_import_prefix(self):
        if self.app.is_user_app:
            # Use the MEERKAT_NPM_PACKAGE package instead of $lib
            # in a Meerkat generated app
            return MEERKAT_NPM_PACKAGE
        return "$lib"

    def render_route(self, page: "Page"):
        template = JINJA_ENV.get_template("page.svelte.jinja")

        # TODO: make this similar to render_root_route
        #       and use component.frontend_alias and component.alias
        return template.render(
            route=page.id,
            title=page.name,
            import_prefix=self.get_import_prefix(),
            components=list(sorted(page.component.get_components())),
            queryparam=False,
        )

    def render_root_route(self):
        template = JINJA_ENV.get_template("page.root.svelte.jinja")
        components = self.get_all_components()
        frontend_components = self.get_all_frontend_components()

        return template.render(
            title="Meerkat",
            import_prefix=self.get_import_prefix(),
            components=components,
            frontend_components=frontend_components,
        )

    def write_component_wrapper(self, component: Type[BaseComponent]):
        cwd = f"{self.appdir}/src/lib/wrappers/__{component.namespace}"
        os.makedirs(cwd, exist_ok=True)
        write_file_if_changed(
            f"{cwd}/{component.__name__}.svelte",
            self.render_component_wrapper(component),
        )

    def write_all_component_wrappers(
        self,
        exclude_classes: Set[str] = {"Component", "BaseComponent"},
    ):
        # Recursively find all subclasses of BaseComponent
        subclasses = get_subclasses_recursive(BaseComponent)
        for subclass in subclasses:
            # Use subclass.__name__ as the component name, instead of
            # subclass.component_name, because the latter is not guaranteed to be
            # unique.
            component_name = subclass.__name__
            if component_name in exclude_classes:
                continue

            # Make a file for the component, inside a subdirectory for the namespace
            # e.g. src/lib/wrappers/__meerkat/Component.svelte
            self.write_component_wrapper(subclass)

    def write_component_context(self):
        write_file_if_changed(
            f"{self.appdir}/src/lib/ComponentContext.svelte",
            self.render_component_context(),
        )


"""
Convert all Python component classes to Svelte component wrappers.

We only run the following code if
  - a script importing `meerkat` is run directly with Python e.g. `python myscript.py`
  - a notebook importing `meerkat` is run directly with Jupyter
  - a script was run with `mk run` and we are in the `mk run` process
  - a script was run with `mk run`, we are in its `uvicorn` subprocess
    and this is a live reload run (i.e. not the first run of the subprocess)
"""
if (
    (not MEERKAT_RUN_PROCESS and not MEERKAT_RUN_SUBPROCESS)
    or MEERKAT_RUN_PROCESS
    or (MEERKAT_RUN_SUBPROCESS and MEERKAT_RUN_RELOAD_COUNT > 1)
) and not MEERKAT_INIT_PROCESS:
    logger.debug("Running SvelteWriter().run().")
    SvelteWriter().run()

if MEERKAT_RUN_SUBPROCESS:
    # Increment the MEERKAT_RUN_RELOAD_COUNT
    # so that the `uvicorn` subprocess knows that it has been reloaded
    # on a subsequent live reload run
    write_file(
        f"{PathHelper().appdir}/.{MEERKAT_RUN_ID}.reload",
        str(MEERKAT_RUN_RELOAD_COUNT + 1),
    )
