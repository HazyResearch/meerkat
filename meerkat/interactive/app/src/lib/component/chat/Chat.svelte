<script lang="ts">
	import { get_rows } from '$lib/api/api';
	import type { DataFrameRef } from '$lib/api/dataframe';
	import { Avatar, Button, Textarea } from 'flowbite-svelte';
	import { createEventDispatcher } from 'svelte';
	import Message from './Message.svelte';
	const eventDispatcher = createEventDispatcher();

	export let df: DataFrameRef;
	export let imgChatbot: string;
	export let imgUser: string;

	$: messages_promise = get_rows(df.ref_id, 0, 100000);

	let value: string = '';
	let send = () => {
		eventDispatcher('send', { message: value });
		value = '';
	};
</script>

<div class="bg-violet-100 p-4 rounded-lg flex flex-col h-full justify-between shadow-md">
	<div class="flex flex-col-reverse overflow-y-scroll h-full">
		{#await messages_promise then messages}
			{#each messages.rows as _, i}
				<Message
					message={messages.get_cell(messages.full_length - i - 1, 'message').data}
					name={messages.get_cell(messages.full_length - i - 1, 'name').data}
					time={messages.get_cell(messages.full_length - i - 1, 'time').data}
				>
					<svelte:fragment slot="avatar">
						<Avatar
							src={messages.get_cell(messages.full_length - i - 1, 'sender').data === 'chatbot'
								? imgChatbot
								: imgUser}
							stacked={true}
							class="mr-2"
						/>
					</svelte:fragment>
				</Message>
			{/each}
		{/await}
	</div>
	<div class="flex mt-4">
		<Textarea
			rows="5"
			placeholder="Your message..."
			class="mx-4 resize-none"
			bind:value
			on:keyup={(e) => {
				e.key === 'Enter' && !e.shiftKey ? send() : null;
			}}
		/>
		<Button class="text-violet-600 dark:text-violet-500" on:click={send} color="light">
			<svg fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-6 h-6">
				<path
					stroke-linecap="round"
					stroke-linejoin="round"
					d="M6 12L3.269 3.126A59.768 59.768 0 0121.485 12 59.77 59.77 0 013.27 20.876L5.999 12zm0 0h7.5"
				/>
			</svg>
		</Button>
	</div>
</div>
