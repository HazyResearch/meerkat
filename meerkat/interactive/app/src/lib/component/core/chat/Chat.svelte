<script lang="ts">
	import { fetchChunk } from '$lib/utils/api';
	import { DataFrameChunk, type DataFrameRef } from '$lib/utils/dataframe';
	import { Avatar, Button, Textarea } from 'flowbite-svelte';
	import { createEventDispatcher } from 'svelte';
	import { writable, type Writable } from 'svelte/store';
	import Message from './Message.svelte';

	const eventDispatcher = createEventDispatcher();

	export let df: DataFrameRef;
	export let imgChatbot: string;
	export let imgUser: string;

	let messages: Writable<DataFrameChunk> = writable(new DataFrameChunk([], [], [], 0, 'pkey'));
	let value: string = '';

	$: fetchChunk({ df: df, start: 0 }).then((newChunk) => messages.set(newChunk));

	let textarea = null;
	// Reset the textarea value when the response is received i.e. df was updated.
	let resetValue = () => {
		value = '';
		if (textarea !== null) {
			// Make the textarea active and editable again.
			textarea.focus();
			textarea.removeAttribute('readonly');
		}
	};
	$: df, resetValue();

	let sendMessage = () => {
		// Remove the newline character at the end of the message if it exists.
		if (value[value.length - 1] === '\n') {
			value = value.slice(0, -1);
		}
		eventDispatcher('send', { message: value });
		if (document.activeElement instanceof HTMLElement) {
			// Deactivate the textarea so that the user can't send the same message again.
			textarea = document.activeElement;
			document.activeElement.blur();
			// Make the textarea non-editable.
			textarea.setAttribute('readonly', 'readonly');
		}
	};
</script>

<div
	class="h-full bg-gray-50 dark:bg-slate-300 p-4 rounded-lg flex flex-col justify-between shadow-md"
>
	<div class="flex flex-col-reverse overflow-y-scroll">
		{#each $messages.rows as _, i}
			<Message
				message={$messages.getCell($messages.fullLength - i - 1, 'message').data}
				name={$messages.getCell($messages.fullLength - i - 1, 'name').data}
				time={$messages.getCell($messages.fullLength - i - 1, 'time').data}
			>
				<svelte:fragment slot="avatar">
					<Avatar
						src={$messages.getCell($messages.fullLength - i - 1, 'sender').data === 'chatbot'
							? imgChatbot
							: imgUser}
						stacked={true}
						class="mr-2"
					/>
				</svelte:fragment>
			</Message>
		{/each}
	</div>
	<div class="flex mt-4">
		<Textarea
			rows="5"
			placeholder="Your message..."
			class="mx-4 resize-none"
			bind:value
			on:keyup={(e) => {
				if (e.key === 'Enter' && !e.shiftKey) {
					sendMessage();
				}
			}}
		/>
		<Button class="text-violet-600 dark:text-violet-500" on:click={sendMessage} color="light">
			<svg fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-6 h-6">
				<path
					stroke-linecap="round"
					stroke-linejoin="round"
					d="M6 12L3.269 3.126A59.768 59.768 0 0121.485 12 59.77 59.77 0 013.27 20.876L5.999 12zm0 0h7.5"
				/>
			</svg>
		</Button>
	</div>
</div>
