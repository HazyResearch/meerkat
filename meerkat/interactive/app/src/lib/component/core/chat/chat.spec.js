import { fireEvent, render, screen } from '@testing-library/svelte';
import { describe, expect, it, vi } from 'vitest';
import Chat from './Chat.svelte';

describe('Chat', () => {
	it('should render chat', async () => {
		const { container, component } = render(Chat, {
			props: {
				df: { refId: 'mock' },
				imgChatbot: 'https://placeimg.com/200/200/animals',
				imgUser: 'https://placeimg.com/200/200/people',
			},
		});

		// Check the messages in the chat are rendered
		expect(await screen.findByText('hello')).toBeTruthy();
		expect(container).toMatchSnapshot();

		// Mock the send event
		const send = vi.fn();
		component.$on(`send`, send);

		// Set the value of the textarea
		const textarea = screen.getByRole('textbox');
		textarea.setAttribute('value', 'blah blah');

		// Click the send button
		const button = screen.getByRole('button');
		await fireEvent.click(button);

		// Check the send event was fired
		expect(send).toHaveBeenCalledTimes(1);
	});
});