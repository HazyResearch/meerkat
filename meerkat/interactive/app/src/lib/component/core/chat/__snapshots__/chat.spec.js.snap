// Vitest Snapshot v1

exports[`Chat > should render chat 1`] = `
<body>
  <div>
    <div
      class="h-full bg-violet-100 p-4 rounded-lg flex flex-col justify-between shadow-md"
    >
      <div
        class="flex flex-col-reverse overflow-y-scroll"
      >
        <div
          class="flex ml-2 pl-4 pr-4 mb-4"
        >
          <img
            alt=""
            class="rounded-full w-10 h-10 border-2 -ml-4 border-white dark:border-gray-800 bg-gray-100 dark:bg-gray-600 text-gray-600 dark:text-gray-300 mr-2"
            src="https://placeimg.com/200/200/people"
          />
          
          <!--&lt;Avatar&gt;-->
           
           
          <div
            class="ml-2 w-full"
          >
            <p
              class="flex justify-between"
            >
              <span
                class="font-bold"
              >
                user
              </span>
               
              <span
                class="text-gray-500"
              >
                2021-01-01 00:00:00
              </span>
            </p>
             
            <p
              class="text-gray-800 whitespace-pre-line"
            >
              im good thanks
            </p>
          </div>
        </div>
        <!--&lt;Message&gt;-->
        <div
          class="flex ml-2 pl-4 pr-4 mb-4"
        >
          <img
            alt=""
            class="rounded-full w-10 h-10 border-2 -ml-4 border-white dark:border-gray-800 bg-gray-100 dark:bg-gray-600 text-gray-600 dark:text-gray-300 mr-2"
            src="https://placeimg.com/200/200/people"
          />
          
          <!--&lt;Avatar&gt;-->
           
           
          <div
            class="ml-2 w-full"
          >
            <p
              class="flex justify-between"
            >
              <span
                class="font-bold"
              >
                chatbot
              </span>
               
              <span
                class="text-gray-500"
              >
                2021-01-01 00:00:00
              </span>
            </p>
             
            <p
              class="text-gray-800 whitespace-pre-line"
            >
              how can i help
            </p>
          </div>
        </div>
        <!--&lt;Message&gt;-->
        <div
          class="flex ml-2 pl-4 pr-4 mb-4"
        >
          <img
            alt=""
            class="rounded-full w-10 h-10 border-2 -ml-4 border-white dark:border-gray-800 bg-gray-100 dark:bg-gray-600 text-gray-600 dark:text-gray-300 mr-2"
            src="https://placeimg.com/200/200/people"
          />
          
          <!--&lt;Avatar&gt;-->
           
           
          <div
            class="ml-2 w-full"
          >
            <p
              class="flex justify-between"
            >
              <span
                class="font-bold"
              >
                user
              </span>
               
              <span
                class="text-gray-500"
              >
                2021-01-01 00:00:00
              </span>
            </p>
             
            <p
              class="text-gray-800 whitespace-pre-line"
            >
              hi
            </p>
          </div>
        </div>
        <!--&lt;Message&gt;-->
        <div
          class="flex ml-2 pl-4 pr-4 mb-4"
        >
          <img
            alt=""
            class="rounded-full w-10 h-10 border-2 -ml-4 border-white dark:border-gray-800 bg-gray-100 dark:bg-gray-600 text-gray-600 dark:text-gray-300 mr-2"
            src="https://placeimg.com/200/200/people"
          />
          
          <!--&lt;Avatar&gt;-->
           
           
          <div
            class="ml-2 w-full"
          >
            <p
              class="flex justify-between"
            >
              <span
                class="font-bold"
              >
                chatbot
              </span>
               
              <span
                class="text-gray-500"
              >
                2021-01-01 00:00:00
              </span>
            </p>
             
            <p
              class="text-gray-800 whitespace-pre-line"
            >
              hello
            </p>
          </div>
        </div>
        <!--&lt;Message&gt;-->
        
      </div>
       
      <div
        class="flex mt-4"
      >
         
        <textarea
          class="w-full rounded-lg bg-gray-50 dark:bg-gray-700 text-gray-900 dark:placeholder-gray-400 dark:text-white  border border-gray-200 dark:border-gray-600 mx-4 resize-none p-2.5 text-sm focus:ring-blue-500 focus:border-blue-500 dark:focus:ring-blue-500 dark:focus:border-blue-500"
          placeholder="Your message..."
          rows="5"
        />
        
        <!--&lt;Wrapper&gt;-->
         
        
        
        <!--&lt;Wrapper&gt;-->
        <!--&lt;Textarea&gt;-->
         
        <button
          class="text-center font-medium focus:ring-4 focus:outline-none inline-flex items-center justify-center px-5 py-2.5 text-sm text-gray-900 bg-white border border-gray-300 hover:bg-gray-100 focus:ring-gray-200 dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:hover:bg-gray-700 dark:hover:border-gray-600 dark:focus:ring-gray-700 rounded-lg text-violet-600 dark:text-violet-500"
          type="button"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 12L3.269 3.126A59.768 59.768 0 0121.485 12 59.77 59.77 0 013.27 20.876L5.999 12zm0 0h7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
        
        <!--&lt;Button&gt;-->
      </div>
    </div>
    <!--&lt;Chat&gt;-->
  </div>
</body>
`;
